### PR TITLE
[amdgpu] Enable struct for on amdgpu backend

### PR DESCRIPTION
**Pull request by GaleSeLee**
Issue: #https://github.com/taichi-dev/taichi/issues/6434

### Brief Summary
1. refactor `add_struct_for_func` into llvm_pass
2. replace `bool spmd` with `string spmd` to recognize `amdgpu` and `nvgpu`.

### CODE REVIEW

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 1]
```diff
@@ -2024,7 +2024,7 @@ std::tuple<llvm::Value *, llvm::Value *> TaskCodeGenLLVM::get_range_for_bounds(
 }
 
 void TaskCodeGenLLVM::create_offload_struct_for(OffloadedStmt *stmt,
-                                                bool spmd) {
+                                                std::string spmd) {
```
Comment (bobcao3) [inline taichi/codegen/llvm/codegen_llvm.cpp:2027]: nit: Use a enum `enum class OffloadSPMDType` or similar

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 2]
```diff
@@ -2126,13 +2126,27 @@ void TaskCodeGenLLVM::create_offload_struct_for(OffloadedStmt *stmt,
...
 
-    if (spmd) {
+    if (spmd == OffloadSPMDType::nvgpu) {
       thread_idx =
           builder->CreateIntrinsic(Intrinsic::nvvm_read_ptx_sreg_tid_x, {}, {});
       block_dim = builder->CreateIntrinsic(Intrinsic::nvvm_read_ptx_sreg_ntid_x,
                                            {}, {});
       builder->CreateStore(builder->CreateAdd(thread_idx, lower_bound),
                            loop_index);
+    } else if (spmd == OffloadSPMDType::amdgpu) {
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:2136]: There's already some caveats in the original codes. Ideally `codegen_llvm.cpp` should have only contain backend-agnostic logics and therefore `spmd` shouldn't appear as an argument to specify GPU-related backends for `create_offload_struct_for()`.

The logic for `spmd` only diverges in terms of `thread_idx` and `block_dim`. One possible way to refactor this is to replace `create_offload_struct_for(..., bool spmd)` with `create_offload_struct_for(..., llvm::Value* thread_idx, llvm::Value* block_dim)`:
- Unify the logic for all backends to:
```
builder->CreateStore(builder->CreateAdd(thread_idx, lower_bound),
                     loop_index);
```
- For AMDGPU/CUDA, implement the backend-specific logic (using Intrinsics) in `codegen_amdgpu.cpp`/`codegen_cuda.cpp`, and then pass in `thread_idx` and `block_dim`
- For other backends, pass `thread_idx = tlctx->get_constant(0)` and `block_dim = tlctx->get_constant(1)`

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 3]
```diff
@@ -2218,7 +2232,7 @@ void TaskCodeGenLLVM::create_offload_struct_for(OffloadedStmt *stmt,
       // body tail: increment loop_index and jump to loop_test
       builder->SetInsertPoint(body_tail_bb);
 
-      if (spmd) {
+      if (spmd == OffloadSPMDType::nvgpu || spmd == OffloadSPMDType::amdgpu) {
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:2235]: Based on the above mentioned refactor, you can remove `spmd` here and put the unified logic:
```
create_increment(loop_index, block_dim);
```

File: taichi/codegen/llvm/codegen_llvm.cpp (hunk 2 above)
Comment (GaleSeLee) [reply taichi/codegen/llvm/codegen_llvm.cpp]: There's an llvm-related error when i implement. `create_offload_struct_for` will create a new llvm::Function named function_body. Thus, `thread_idx` and `block_dim` which we pass through `create_offload_struct_for(..., llvm::Value* thread_idx, llvm::Value* block_dim)` will can't be used in `function_body`. :-)
More detail:
```
define internal void @function_body(%struct.RuntimeContext* %0, i8* %1, %struct.Element* %2, i32 %3, i32 %4) {
allocs:
  %5 = alloca i32, align 4
  %6 = alloca %struct.PhysicalCoordinates, align 8
  %7 = alloca %struct.PhysicalCoordinates, align 8
  %8 = alloca %struct.DenseMeta, align 8
  br label %entry

final:                                            ; preds = %func_exit
  ret void

entry:                                            ; preds = %allocs
  br label %function_body

function_body:                                    ; preds = %entry
  %9 = call %struct.PhysicalCoordinates.0* @Element_get_ptr_pcoord(%struct.Element* %2)
  %10 = bitcast %struct.PhysicalCoordinates.0* %9 to %struct.PhysicalCoordinates*
  call void @S3_refine_coordinates(%struct.PhysicalCoordinates* %10, %struct.PhysicalCoordinates* %6, i32 0)
  %11 = getelementptr i8, i8* %1, i64 0
  %12 = bitcast i8* %11 to i32*
  store i32 0, i32* %12, align 4
  %13 = add i32 %0, %3
// Actually, this %0 is from %0 = call i32 @llvm.amdgcn.workitem.id.x() not %struct.RuntimeContext* %0
```

Comment (GaleSeLee) [reply taichi/codegen/llvm/codegen_llvm.cpp]: The implementation uses the virtual func `get_spmd_info` to get `thread_idx` and `block_dim`.

**Pull request merged**